Feature/refactoring

peterchoee opened the pull request:
# FEATURE PR

## 설명
- app에 있던 presentation layer의 코드를 presentation 모듈 생성 후 이전
- data module, domain module 패키지에 artifact을 포함하도록 변경
- data module(layer) 내 Mapper, Entity 생성 & 패키지 구조 변경
- domain module(layer) 내 Entity, UseCase 생성
- presentation module(layer) 내 Mapper, Entity 생성 & UI 작업 진행

## PR 타입
- [ ] 새 기능
- [ ] 큰 변경사항
- [x] 코드스타일 수정
- [x] 리펙터링
- [ ] 문서내용 변경
- [ ] 그외 기타

Commit:
[Upate] app 모듈에서 presentation 모듈 분리, Main UI

기존의 app 모듈 내의 presentation layer 코드를 presentation module로 분리
MainActivity, ViewModel, BindingAdapter, BaseRecyclerAdapter, ViewHolder 작업 진행

Commit:
[Update] 모듈별 Gradle 정리

Commit:
[Add&Update] Data Layer Entity, Mapper, Classes

모듈 package, artifact 수정 (com.ddd4.data -> com.ddd4.dropit.data)
Mapper Extension 생성
utils 코드를 presentation 모듈로 이전
SharedPreferencesHelper를 LocalDataSource에서 처리하도록 변경

Commit:
[Add&Update] Domain Layer Entity, Mapper, Classes

모듈 package, artifact 수정 (com.ddd4.domain -> com.ddd4.dropit.domain)
Mapper Extension 생성
SharedPreferencesRepository 삭제
Folder(폴더), Item(아이템)에 대한 UseCase 구현

Commit:
[Fix] DI 모듈 수정

모듈 컴포넌트 변경
변경된 Data, Domain에 따라 Provide 수정

mkw8263 commented on presentation/src/main/java/com/ddd4/dropit/presentation/ui/main/MainViewModel.kt:
```diff
@@ -0,0 +1,49 @@
+
+    private fun getFolderItems() {
+        viewModelScope.launch {
+            getFolderUseCase.execute().also { result ->
```
여기서 쓰는 standard function 중 also를 쓰셨는데 요부분은 also 가 아니라 run이 맞는거같아요!
also 쓰이는 케이스를 한번 찾아보시면 좋을꺼같아요:)

peterchoee replied:
@mkw8263 
👍 아직 스코프 함수가 조금 헷갈려서 ㅎㅎ

mkw8263 commented on data/src/main/java/com/ddd4/dropit/data/source/local/LocalDataSourceImpl.kt:
```diff
@@ -0,0 +1,22 @@
+): LocalDataSource {
+
+    override suspend fun getFolders(): List<DomainEntity.Folder> =
+        databaseDao.selectFolders().map { data ->
```
요 부분은 .map{ } 을 굳이 안써도 될 부분인거같아요!

peterchoee replied:
@mkw8263 
오! 그러네요! ㅎㅎ 감사합니당

jsh-me commented on presentation/src/main/java/com/ddd4/dropit/presentation/util/ImageViewUtil.kt:
```diff
@@ -1,33 +1,34 @@
 fun ImageView.loadUriCenterCrop(uri : Uri) {
     val options = RequestOptions()
-    GlideApp.with(this).load(uri).apply(options.centerCrop()).into(this)
+    Glide.with(this).load(uri).apply(options.centerCrop()).into(this)
```
https://bumptech.github.io/glide/doc/configuration.html#glidemodule 을 참고하여 만든 util 입니다. Glide.with()..... 로 진행하실거면 DropItGlideApp.kt를 삭제하는게 좋을 것 같아요!!

peterchoee replied:
@jsh-me 
아니에요! 이거 제가 잘 못 이해한거 같아요!
참고해서 다시 수정할게요 👍

jsh-me reviewed (approved):
굳굳굳굳

TaeseongYun commented on data/src/main/java/com/ddd4/dropit/data/source/local/room/DatabaseDao.kt:
```diff
@@ -0,0 +1,32 @@
+    @Insert(onConflict = OnConflictStrategy.REPLACE)
+    suspend fun insertFolder(folderData: DataEntity.Folder)
+
+    @Delete
```
함수명이 update로 시작하던데 Update 말고 Delete로 주신이유가 궁금합니다 ~

jsh-me replied:
@TaeseongYun 얼 날카로우셔 대단쓰

peterchoee replied:
@TaeseongYun 
엇! 실수로 못고쳤나봐요! @Update 맞습니다!

TaeseongYun reviewed (commented):
궁금한점을 코멘트로 남겼습니다 ~

mkw8263 commented on domain/src/main/java/com/ddd4/dropit/domain/usecase/getFolderUseCase.kt:
```diff
@@ -0,0 +1,24 @@
+            Result.Error(e)
+        }
+
+    sealed class Result {
```
Result<T> 를 지금 매 usecase 마다 정의해서 쓰고 있는데 제네릭하게 하나 만들면 usecase 쪽 코드가 더 좋아질꺼 같습니다

peterchoee replied:
@mkw8263 
좋은 의견 감사해요 👍 
경운님 의견대로 제네릭하게 만들어서 같이 사용해야겠어요!

mkw8263 commented on presentation/src/main/java/com/ddd4/dropit/presentation/ui/main/MainActivity.kt:
```diff
@@ -0,0 +1,34 @@
+        })
+        mainViewModel.startCategory.observe(this, Observer { categoryId ->
+            startActivity(Intent(this, CategoryActivity::class.java)
+                .putExtra("categoryId", categoryId))
```
extra key 값의 string 을 하드 코딩말고 실행할 Activity class의 companion object 정리해서 key 로 쓰시는게 좋아보여요:)
``` kotlin
class CategoryActivity : BaseActivity~~~{
   companion object {
     const val EXTRA_KEY_CATEGORY_ID = "키키키키키키키키"
   }
}
```

peterchoee replied:
@mkw8263 
👍

Commit:
Fix. UseCase 클래스명, Result 클래스 분리

각 UseCase 명칭이 파스칼케이스에 맞지 않아서 대문자로 수정, DI Module 변경
UseCase 내에 공통적으로 있던 Result 클래스 분리

Commit:
Fix. DatabaseDao annotation

Update 함수에 잘못적용되어 있던 어노테이션 수정 (@Delete -> @Update)

Commit:
Fix. mapper 함수참조 형식 변경

Commit:
Fix. Glide Base

Commit:
Fix. MainActivity Extra name

Extra name을 상수로 변경

Commit:
Fix. Scope Function 수정

Scope Function을 리시버를 반환하도록 변경 (also -> run)

jsh-me committed:
Fix conflict

jsh-me commented:
정렬님짱~

jsh-me commented on presentation/src/main/java/com/ddd4/dropit/presentation/ui/main/MainViewModel.kt:
```diff
@@ -0,0 +1,50 @@
+                when (this) {
+                    is Result.Success -> {
+                        if (data.isNotEmpty()) {
+                            _folderItems.value = data.map(DomainEntity.Folder::mapToPresentation)
```
@choedeb 
머지하고 확인했네요 ㅠ 제 불찰입니다 지송지송
우리 저번에 얘기했던거 기억나나여?
이건 뷰에 대한 변화가 아닌 데이터의 변화니까 뷰모델에서 로직처리를 최대한 배제하고, 유스케이스에서 해당 로직을 진행해보는게 어떨까요?
_folderItems.value = getFolderUseCase() 로 한방에 끝내버리기! 당신의 생각은?

jsh-me replied:
https://github.com/android/architecture-samples/blob/usecases/app/src/main/java/com/example/android/architecture/blueprints/todoapp/domain/GetTasksUseCase.kt

이런식으로 하면 괜찮지 않을까~ 생각해봤습니당

peterchoee replied:
@jsh-me 
엇.. ㅎㅎ 확인해볼게요!!
